src/rust/engine/fs/store/src/lib.rs

```diff
@@ -238,29 +238,51 @@ impl RemoteStore {
 
   /// Download the digest to the local byte store from this remote store. The function `f_remote`
   /// can be used to validate the bytes.
-  async fn download_digest_to_local<
-    FRemote: Fn(Bytes) -> Result<(), String> + Send + Sync + 'static,
-  >(
+  async fn download_digest_to_local(
     &self,
     local_store: local::ByteStore,
     digest: Digest,
     entry_type: EntryType,
-    f_remote: FRemote,
+    f_remote: Option<&(dyn Fn(Bytes) -> Result<(), String> + Send + Sync + 'static)>,
   ) -> Result<(), StoreError> {
     let remote_store = self.store.clone();
     self
       .maybe_download(digest, async move {
-        let bytes = remote_store.load_bytes(digest).await?.ok_or_else(|| {
-          StoreError::MissingDigest(
-            "Was not present in either the local or remote store".to_owned(),
-            digest,
-          )
-        })?;
+        // if there's a function to call, always just buffer fully into memory
+        let data = remote_store
+          .load(digest, f_remote.is_some())
+          .await?
+          .ok_or_else(|| {
+            StoreError::MissingDigest(
+              "Was not present in either the local or remote store".to_owned(),
+              digest,
+            )
+          })?;
 
-        f_remote(bytes.clone())?;
-        let stored_digest = local_store
-          .store_bytes(entry_type, None, bytes, true)
-          .await?;
+        if let Some(f_remote) = f_remote {
+          // if there's a function, just slurp the contents back into memory
+          match data {
+            Either::Left(ref bytes) => f_remote(bytes.clone())?,
+            Either::Right(_) => panic!("unexpectedly got file output when using f_remote"),
```

Comment:
```suggestion
            Either::Right(_) => unreachable!("unexpectedly got file output when using f_remote"),
```
... to send the signal that this is supposed to be an impossible case.